src/java/htsjdk/samtools/SAMRecord.java

```diff
@@ -305,119 +308,202 @@ private boolean hasMateReferenceName() {
+            }
         }
         return mReferenceIndex;
     }
 
     /**
-     * @param referenceIndex Must either equal -1 (indicating no reference), or exist in the sequence dictionary
-     * in the header associated with this record.
+     * Updates the reference index. The record must have a valid SAMFileHeader, and the reference index must appear in
+     * the header's sequence dictionary. If the reference index is valid, the reference name will also be resolved
+     * and updated to the name for the sequence dictionary entry corresponding to the index.
+     *
+     * @param referenceIndex Must either equal NO_ALIGNMENT_REFERENCE_INDEX (-1) indicating no reference, or the
+     *                       record must have a SAMFileHeader and the index must exist in the associated sequence
+     *                       dictionary.
+     * @throws IllegalArgumentException if the SAMFileHeader is null for this record or the reference index is not
+     *         found in the sequence dictionary for this record.
      */
     public void setReferenceIndex(final int referenceIndex) {
-        mReferenceIndex = referenceIndex;
-        if (mReferenceIndex == NO_ALIGNMENT_REFERENCE_INDEX) {
+        if (referenceIndex == NO_ALIGNMENT_REFERENCE_INDEX) {
+            mReferenceIndex = NO_ALIGNMENT_REFERENCE_INDEX;
             mReferenceName = NO_ALIGNMENT_REFERENCE_NAME;
-        } else {
-            try {
-                mReferenceName = mHeader.getSequence(referenceIndex).getSequenceName();
-            } catch (final NullPointerException e) {
-                throw new IllegalArgumentException("Reference index " + referenceIndex + " not found in sequence dictionary.", e);
+        }
+        else if (null == mHeader) {
+            throw new IllegalArgumentException("A non-null SAMFileHeader is required to resolve the reference index");
+        }
+        else {
+            SAMSequenceRecord samSequence = mHeader.getSequence(referenceIndex);
+            if (null != samSequence) {
+                mReferenceIndex = referenceIndex;
+                mReferenceName = samSequence.getSequenceName();
+            }
+            else {
+                throw new IllegalArgumentException("Reference index " + referenceIndex + " not found in sequence dictionary.");
             }
         }
     }
 
     /**
-     * @return Mate reference name, or null if one is not assigned.
+     * @return Reference name, or NO_ALIGNMENT_REFERENCE_NAME (-1) if the record has no reference name
      */
     public String getMateReferenceName() {
         return mMateReferenceName;
     }
 
+    /**
+     * Sets the mate reference name for this record. If the record has a valid SAMFileHeader and the mate reference
+     * name is present in the associated sequence dictionary, the record's mate reference index will also be
+     * updated with the corresponding sequence index.
```

Comment:
"If mateReferenceName is NO_ALIGNMENT_REFERENCE_NAME, sets the mate reference index to NO_ALIGNMENT_REFERENCE_INDEX"